docs/topics/async.txt

```diff
@@ -4,18 +4,110 @@ Asynchronous support
+still work under WSGI, but with performance penalties and without the ability
+to have long-running requests efficiently.
 
-There is limited support for other parts of the async ecosystem; namely, Django
-can natively talk :doc:`ASGI </howto/deployment/asgi/index>`, and some async
-safety support.
+We're still working on asynchronous support for the ORM and other parts of
+Django; you can expect to see these in future releases. For now, you can use
+the ``sync_to_async`` adapter to interact with normal Django, as well as use a
+whole range of Python asyncio libraries natively. See below for more details.
+
+.. versionchanged:: 3.1
+
+    Support for async views was added.
+
+Async views
+===========
+
+.. versionadded:: 3.1
+
+Any view can be declared async by making the callable part of it return a
+coroutine - commonly, this is done using ``async def``. For a function-based
+view, this simply means declaring the whole view using ``async def``; for a
+class-based view, this means making its ``__call__`` method an ``async def``
+(not its ``__init__`` or ``as_view``!).
+
+.. note::
+
+    Django uses ``asyncio.iscoroutinefunction`` to test if your view is
+    asynchronous or not. If you implement your own method of returning a
+    coroutine, ensure you set the ``_is_coroutine`` attribute of the view
+    to ``asyncio.coroutines._is_coroutine`` so this function returns ``True``.
+
+Under a WSGI server, asynchronous views will run in their own, one-off event
+loop; this means that you can do things like parallel, async HTTP calls to APIs
+without any issues, but you will not get the benefits of an asynchronous request
+stack.
+
+If you want these benefits - which are mostly around the ability to service
+hundreds of connections without using any Python threads (enabling slow
+streaming, long-polling and other exciting response types) - you will need to
+deploy Django using :doc:`ASGI </howto/deployment/asgi/index>` instead.
+
+.. warning::
+
+    You will only get the benefits of a fully-asynchronous request stack if you
+    have *no synchronous middleware* loaded into your site; if there is a piece
+    of synchronous middleware, then Django must use a thread per request to
+    safely emulate a synchronous environment for it.
+
+    Middleware can be built to support
+    :ref:`both sync and async <async-middleware>` contexts; some of Django's
+    middleware is built like this, but not all. To see what middleware Django
+    has to adapt, you can turn on debug logging and look for log messages
+    about "adapting synchronous middleware".
+
+In either ASGI or WSGI mode, though, you can safely use asynchronous support
+to run code in parallel rather than serially, which is especially handy when
+dealing with external APIs or datastores.
+
+If you want to call a part of Django that is still synchronous (like the ORM)
+you will need to wrap it in a ``sync_to_async`` call, like this::
```

Comment:
I feel there's a link to the `asgiref` docs missing somewhere — it's not really discussed. 

(Can we do Sphinx links for those...? 🤔)

Reply:
I've omitted linking to the asgiref docs purely because there's not much there of substance; async_to_sync and sync_to_async are documented more in the Django codebase than in asgiref (which is more general-ASGI docs).

Reply:
I have a lingering feeling they need some documentation. Maybe just a section here, saying what they are?

Reply:
Yup, I'll write up some reference docs quickly and put them at the bottom there for now.

Reply:
OK, docs added. They're a little long but I feel that's better than being too short.

Reply:
Super. Thanks @andrewgodwin. I'll give it a look tomorrow now. Have a good rest of your day!